[ja] add translation of content/en/docs/languages/dotnet/logs/best-practices.md (#10799)
open-telemetry/opentelemetry.io: content/ja/docs/languages/dotnet/logs/best-practices.md

#### パッケージバージョン {#package-version}

使用している .NET ランタイムのバージョンに関係なく、[Microsoft.Extensions.Logging](https://www.nuget.org/packages/Microsoft.Extensions.Logging/) パッケージの最新の安定バージョンの [`ILogger`](https://docs.microsoft.com/dotnet/api/microsoft.extensions.logging.ilogger) インターフェイス（[`ILogger<TCategoryName>`](https://learn.microsoft.com/dotnet/api/microsoft.extensions.logging.ilogger-1) を含む）を使用してください。

- [OpenTelemetry .NET SDK](/docs/languages/dotnet/) の最新の安定バージョンを使用している場合、`Microsoft.Extensions.Logging` パッケージのバージョンはパッケージの依存関係を通じてすでに管理されているため、気にする必要はありません。
- バージョン `3.1.0` 以降、.NET ランタイムチームはメジャーバージョンの更新時でも `Microsoft.Extensions.Logging` の後方互換性に高い基準を設けているため、互換性について心配する必要はありません。

#### ロガーの取得 {#get-logger}

`ILogger` インターフェイスを使用するには、まずロガーを取得する必要があります。
ロガーの取得方法は次の2つの要素に依存します。

- 構築しているアプリケーションの種類。
- ログを記録したい場所。

一般的なルールとして、以下のようになります。

- [依存性注入（DI）](https://learn.microsoft.com/dotnet/core/extensions/dependency-injection)を使用するアプリケーション（たとえば [ASP.NET Core](https://learn.microsoft.com/aspnet/core) や [.NET Worker](https://learn.microsoft.com/dotnet/core/extensions/workers)）を構築している場合、ほとんどのケースでは DI が提供するロガーを使用するべきですが、DI のロギングパイプラインが利用可能になる前や、破棄された後にログを記録したい特殊なケースもあります。

>   詳しくは [.NET 公式ドキュメント](https://learn.microsoft.com/dotnet/core/extensions/logging#integration-with-hosts-and-dependency-injection)と [OpenTelemetry .NET ログを5分で始める - ASP.NET Core アプリケーション](/docs/languages/dotnet/logs/getting-started-aspnetcore/)チュートリアルを参照してください。

Reviewer: ```suggestion
  詳しくは [.NET 公式ドキュメント](https://learn.microsoft.com/dotnet/core/extensions/logging#integration-with-hosts-and-dependency-injection)と [5分で始める OpenTelemetry .NET ログ - ASP.NET Core アプリケーション](/docs/languages/dotnet/logs/getting-started-aspnetcore/)チュートリアルを参照してください。
```

[メトリクスのベストプラクティスページ](https://github.com/open-telemetry/opentelemetry.io/blob/main/content/ja/docs/languages/dotnet/metrics/best-practices.md?plain=1#L160)に合わせた訳のほうが自然に感じました｡

- DI を使用しないアプリケーションを構築している場合は、[LoggerFactory](#loggerfactory) インスタンスを作成し、OpenTelemetry がそれと連携するように設定してください。

>   詳しくは [OpenTelemetry .NET ログを5分で始める - コンソールアプリケーション](/docs/languages/dotnet/logs/getting-started-console/)チュートリアルを参照してください。

Reviewer: ```suggestion
  詳しくは [5分で始める OpenTelemetry .NET ログ - コンソールアプリケーション](/docs/languages/dotnet/logs/getting-started-console/)チュートリアルを参照してください。
```

こちらの訳も[メトリクスのベストプラクティスページ](https://github.com/open-telemetry/opentelemetry.io/blob/main/content/ja/docs/languages/dotnet/metrics/best-practices.md?plain=1#L162)に合わせたほうが自然に感じました｡

たとえば、以下のコードではロガーが有効でない場合、`Database.GetFoodPrice` の呼び出しはスキップされます。

```csharp
if (logger.IsEnabled(LogLevel.Information))
{
    logger.SayHello(food, Database.GetFoodPrice(food));
}
```

上記のシナリオでは `IsEnabled` がパフォーマンス上の利点をもたらす場合がありますが、ほとんどのユーザーにとってはより多くの問題を引き起こす可能性があります。
たとえば、コードのパフォーマンスがどのロガーが有効かに依存するようになり、さらに引数の評価にはロギング設定に依存する重大な副作用がある可能性もあります。

コンパイル時ソースジェネレーターを使用する場合は、例外をログに記録するための専用パラメーターを使用してください。
たとえば、以下のようになります。

```csharp
var food = "tomato";
var price = 2.99;

try
{
    // ロジックを実行

    logger.SayHello(food, price);
}
catch (Exception ex)
{
    logger.SayHelloFailure(ex, food, price);
}

internal static partial class LoggerExtensions
{
    [LoggerMessage(Level = LogLevel.Information, Message = "Hello from {food} {price}.")]
    public static partial void SayHello(this ILogger logger, string food, double price);

    [LoggerMessage(Level = LogLevel.Error, Message = "Could not say hello from {food} {price}.")]
    public static partial void SayHelloFailure(this ILogger logger, Exception exception, string food, double price);
}
```

> [!NOTE]
>
> コンパイル時ソースジェネレーターを使用する場合、検出された最初の `Exception` パラメーターには自動的に特別な処理が適用されます。
> これはメッセージテンプレートの一部にすべきでは**ありません**。

> > 詳しくは [Log method anatomy](https://learn.microsoft.com/dotnet/core/extensions/logger-message-generator#log-method-anatomy) を参照してください。

Reviewer: ```suggestion
> 詳しくは [ログメソッドの構造](https://learn.microsoft.com/dotnet/core/extensions/logger-message-generator#log-method-anatomy) を参照してください。
```

リンク先の日本語ページに合わせて､ ｢ログメソッドの構造｣ でも良いと思いましたが､ いかがでしょうか｡
